added service implementation 101 doc

danbugs opened the pull request:
Signed-off-by: Dan Chiarlone <dchiarlone@microsoft.com>

devigned commented on docs/service-implementation-101.md:
```diff
@@ -0,0 +1,390 @@
+
+> Note: If you would like to develop a new capability outside of our provided interfaces, you will first have to create its' WIT. Please, create an issue to start a design discussion, and make a PR with the `.wit` file itself prior to its' `slight` implementation.
+
+The DeisLabs Engineering Team has designed each of its' main capabilities to be easily extensible, and some macros to take care of boiler-plate code for you. In this tutorial, we will implement a new `pubsub` implementation — a local one with Mosquitto!
```
```suggestion
The DeisLabs Engineering Team has designed each of its' capabilities to be easily extensible, and some macros to take care of boilerplate code for you. In this tutorial, we will implement a new `pubsub` implementation — a local one with Mosquitto!
```

devigned commented on docs/service-implementation-101.md:
````diff
@@ -0,0 +1,390 @@
+Assuming that you are on the root of the SpiderLightning repository that you locally cloned, to get started, do:
+
+```sh
+git branch my_local_pubsub_implementation c73826a1d522bc3e1d2f0387481880edecd3e3d3
````
```suggestion
git checkout -b my_local_pubsub_implementation
```

danbugs replied:
I want this tutorial to work regardless of the state of the main branch~ That's why I'm branching off of a specific commit hash.

devigned replied:
Won't that create a large PR when targeting main, or cause the developer to need to rebase to get started?

devigned commented on docs/service-implementation-101.md:
```diff
@@ -0,0 +1,390 @@
+
+## Getting Started
+
+Assuming that you are on the root of the SpiderLightning repository that you locally cloned, to get started, do:
```
```suggestion
Assuming that you are on the root of the SpiderLightning repository that you locally cloned, to get started, create a new branch:
```

devigned commented on docs/service-implementation-101.md:
```diff
@@ -0,0 +1,390 @@
+
+Are you hoping to learn how to implement a new service for our existing capabilities (i.e., kv, lockd, mq, or pubsub)? You have come to the right place!
+
+> Note: If you would like to develop a new capability outside of our provided interfaces, you will first have to create its' WIT. Please, create an issue to start a design discussion, and make a PR with the `.wit` file itself prior to its' `slight` implementation.
```
```suggestion
> Note: If you would like to develop a new capability outside of our provided interfaces, you will first have to create its WIT (WebAssembly Interface Types) file. Please, create an issue to start a design discussion, and make a PR with the `.wit` file itself prior to its' `slight` implementation.
```

devigned commented on docs/service-implementation-101.md:
```diff
@@ -0,0 +1,390 @@
+
+> Note: Our `Local` variants for both `enum`s hold a struct called `MosquittoImplementor` but we haven't implemented that yet.
+
+With that done, we now have to take care of the `new` function for both `PubImplementor`, and `SubImplementor`. Considering that the `new` function is what is responsible for mapping a string from a user's slightfile to a specific implementor, this is the time where we decide what string should refer to our local implementation — let's say `pubsub.mosquitto`. Now, we can go ahead and change the ` ...
```
```suggestion
With that done, we now have to take care of the `new` function for both `PubImplementor`, and `SubImplementor`. Considering that the `new` function is what is responsible for mapping a string from a user's slightfile to a specific implementor, this is the time where we decide what string should refer to our local implementation — let's say `pubsub.mosquitto`. Now, we can change the `new` functions, like so:
```

devigned commented on docs/service-implementation-101.md:
````diff
@@ -0,0 +1,390 @@
+    }
+```
+
+In here, all we are doing is adding calls to the functions implemented by the `MosquittoImplementor` under the `Local` variant of the `Pub/SubImplementor` enums, so that we are handling the case where a user provides a `toml` with `pubsub.mosquitto` and makes use of its' functionality.
````
```suggestion
In here, all we are doing is adding calls to the functions implemented by the `MosquittoImplementor` under the `Local` variant of the `Pub/SubImplementor` enums, so that we are handling the case where a user provides a `toml` with `pubsub.mosquitto` and makes use of its functionality.
```

devigned commented on docs/service-implementation-101.md:
````diff
@@ -0,0 +1,390 @@
+To finish off, make sure to add your example in the list of example runs under the `run-rust` Makefile rule — for our new pubsub implementor, it will look like so:
+```Makefile
+    RUST_LOG=$(LOG_LEVEL) $(SLIGHT) -c './examples/pubsub-consumer-demo/mosquitto_slightfile.toml' run -m ./examples/pubsub-consumer-demo/target/wasm32-wasi/release/pubsub-consumer-demo.wasm &
+	RUST_LOG=$(LOG_LEVEL) $(SLIGHT) -c './examples/pubsub-producer-demo/mosquitto_slightfile.toml' run -m ./examples/pubsub-producer-demo/target/wasm32-wasi/release/pubsub-producer-demo.wasm
````
nit: indentation here is off.

danbugs replied:
fixed ✅

devigned commented on docs/service-implementation-101.md:
````diff
@@ -0,0 +1,390 @@
+	RUST_LOG=$(LOG_LEVEL) $(SLIGHT) -c './examples/pubsub-producer-demo/mosquitto_slightfile.toml' run -m ./examples/pubsub-producer-demo/target/wasm32-wasi/release/pubsub-producer-demo.wasm
+```
+
+With this, you should have the basic understanding on how to get started developing capabilities for Slight — Now, all you are missing is just the implementation of the service itself! We are looking forward to your contributions 🚀
````
```suggestion
With this, you should have the basic understanding on how to get started developing capabilities for Slight. Now, all you are missing is just the implementation of the service itself! We are looking forward to your contributions 🚀
```

devigned reviewed (changes requested):
Solid overview of creating a service. I've left some feedback on a handful of little items. 

Do you intend on linking to the docs from the main readme?

danbugs commented:
@devigned ~ Thank you for your review! I've committed all your suggestions, fixed the indentation for the `Makefile` rule, and added a link to the `service-implementation-101.md` doc in the main `README.md`!

devigned commented on docs/service-implementation-101.md:
```diff
@@ -0,0 +1,389 @@
+
+> Note: If you would like to develop a new capability outside of our provided interfaces, you will first have to create its WIT (WebAssembly Interface Types) file. Please, create an issue to start a design discussion, and make a PR with the `.wit` file itself prior to its' `slight` implementation.
+
+The DeisLabs Engineering Team has designed each of its' capabilities to be easily extensible, and some macros to take care of boilerplate code for you. In this tutorial, we will implement a new `pubsub` implementation — a local one with Mosquitto!
```
```suggestion
The DeisLabs Engineering Team has designed each of its capabilities to be easily extensible, and some macros to take care of boilerplate code for you. In this tutorial, we will implement a new `pubsub` implementation — a local one with Mosquitto!
```

devigned reviewed (changes requested):
one more its'

danbugs commented:
Thank you for finding all my "its'", @devigned 😆~ fixed!

devigned reviewed (approved):
LGTM! Nice work!
